src/cmd/ksh93/sh/init.c

```diff
@@ -910,52 +910,56 @@ static_fn char *get_version(Namval_t *np, Namfun_t *fp) { return (nv_getv(np, fp
 static Sfdouble_t nget_version(Namval_t *np, Namfun_t *fp) {
     UNUSED(np);
     char *cp = strdup(SH_RELEASE);
-    Sflong_t t = 0;
+    // We should not need to convert version number string every time this function is called
+    static Sflong_t version_number  = -1;
     char *dash;
     char *dot;
     char *major_str, *minor_str, *patch_str;
     int major, minor, patch;
     UNUSED(fp);
 
-    // Version string in development version could be set like 2017.0.0-devel-1509-g95d59865
-    // If a '-' exists in version string, set it as end of string i.e. version string becomes
-    // 2017.0.0
-    dash = strchr(cp, '-');
-    if (dash) *dash = 0;
-
-    // Major version number starts at beginning of string
-    major_str = cp;
-
-    // Find the first '.' and set it to NULL, so major version string is set to 2017
-    dot = strchr(cp, '.');
-    if (!dot) {
-        // If there is no . in version string, it means version string is either empty, invalid
-        // or it's using old versioning scheme.
-        free(cp);
-        return 0;
-    }
+    if (version_number == -1) {
```

Comment:
Personally, I think an early return is cleaner:

```
if (version_number != -1) return version_number;
```

Then 99% of the function no longer needs to be an indented block.

Reply:
Also, if you put this at the top of the function with the other vars defined below it then you avoid the need to pointlessly `strdup(SH_RELEASE)` and free the buffer even though we've already calculated and cached the return value.